<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px"> 

## Lab: SQL Practice

_Author: Matt Brems_

### Required Preparation

In order to run SQL queries within Python/this Jupyter notebook, you will need to install the following:

```bash
conda install sqlalchemy
conda install psycopg2
```

If you get an error trying to install `psycopg2`, you can install a stand-alone package (sufficient for everything you'll need to complete this lab) by entering the following anywhere in the terminal: 

```bash 
pip install psycopg2-binary
```

(If `pip` doesn't work for you, try `pip3 install psycopg2-binary`.)

In [2]:
conda install sqlalchemy


Note: you may need to restart the kernel to use updated packages.
Solving environment: ...working... done

## Package Plan ##

  environment location: C:\Users\Tresha\anaconda3

  added / updated specs:
    - sqlalchemy


The following packages will be downloaded:


    package                    |            build
    ---------------------------|-----------------
    conda-4.10.1               |   py38haa95532_1         2.9 MB
    ------------------------------------------------------------
                                           Total:         2.9 MB

The following packages will be UPDATED:

  conda              conda-forge::conda-4.10.1-py38haa244f~ --> pkgs/main::conda-4.10.1-py38haa95532_1




conda-4.10.1         | 2.9 MB    |            |   0% 
conda-4.10.1         | 2.9 MB    |            |   1% 
conda-4.10.1         | 2.9 MB    | #4         |  15% 
conda-4.10.1         | 2.9 MB    | ###2       |  32% 
conda-4.10.1         | 2.9 MB    | ####8      |  48% 
conda-4.10.1       

In [4]:
conda install psycopg2-binary

Note: you may need to restart the kernel to use updated packages.



PackagesNotFoundError: The following packages are not available from current channels:

  - psycopg2-binary

Current channels:

  - https://repo.anaconda.com/pkgs/main/win-64
  - https://repo.anaconda.com/pkgs/main/noarch
  - https://repo.anaconda.com/pkgs/r/win-64
  - https://repo.anaconda.com/pkgs/r/noarch
  - https://repo.anaconda.com/pkgs/msys2/win-64
  - https://repo.anaconda.com/pkgs/msys2/noarch

To search for alternate channels that may provide the conda package you're
looking for, navigate to

    https://anaconda.org

and use the search bar at the top of the page.





Solving environment: ...working... failed with initial frozen solve. Retrying with flexible solve.
Solving environment: ...working... failed with initial frozen solve. Retrying with flexible solve.


### Instructions

In this lab, you're going to have ten prompts. Each prompt will require you to do two things:
1. Write a SQL query.
2. Use the result of that query to answer the question.

The purpose of this lab is to get you to practice your SQL - an **important** skill in data science! While it is possible to get these answers using Pandas, it should be your goal to do everything only in SQL.

**Data**: The data used in this lab is the Iowa liquor database. It contains various tables of transactions. The schema for this database is here:

<img src="./images/schema.png" alt="schema" width="750"/>

In [5]:
pip install psycopg2-binary

In [6]:
# Import Pandas and Create_Engine

import pandas as pd
from sqlalchemy import create_engine

# Set up the engine to access the data.
engine = create_engine('postgresql://analytics_student:analyticsga@analyticsga-psql.generalassemb.ly:5432/iowa_liquor_sales_database')

### Prompt 1

How many rows are there in the `stores` table of the Iowa liquor database?

In [9]:
# Use a SQL query to find the answer to the above prompt.

sql = """    

SELECT COUNT(*)
FROM STORES
"""

pd.read_sql_query(sql, engine)

count
0   1973

**Prompt 1 Answer**: 1973

### Prompt 2

If you sort the stores in alphabetical order by name, which store is fourth?

In [14]:
# Use a SQL query to find the answer to the above prompt.

sql = """

SELECT *
FROM stores
ORDER BY name ASC

"""

pd.read_sql_query(sql, engine)

store                                name store_status  \
0      4986  218 Fuel Express & Chubby's Liquor            A   
1      3047          3047 Cub Foods / Iowa City            I   
2      3060               3060 Cub Foods / Ames            I   
3      3061         3061 Cub Foods / Sioux City            I   
4      4845                  3rd St Convenience            A   
...     ...                                 ...          ...   
1968   3992       Xo Food And Liquor / Waterloo            I   
1969   4066                       Xpress Liquor            I   
1970   4886                      Yr Dollar Stop            A   
1971   4908                Zapf's Pronto Market            A   
1972   4874                      Z's Quickbreak            A   

                                          store_address address_info  
0     68 Monroe St\nFloyd, IA 50435\n(43.12249790500...         None  
1     855 Highway 1 West\nIowa City, IA 522460000\n(...         None  
2     3121 Grand Ave\nAmes, IA 500100000\n(42.053458...         None  
3     1732 Hamilton Blvd\nSioux City, IA 511030000\n...         None  
4     216, 3rd St Se\nCedar Rapids, IA 52401\n(41.97...         None  
...                                                 ...          ...  
1968  428 Franklin St\nWaterloo, IA 507030000\n(42.5...         None  
1969  1201 North Jefferson Ste 100\nIndianola, IA 50...         None  
1970  1500 1st Ave Ne\nCedar Rapids, IA 52402\n(41.9...         None  
1971  107 1st St Sw\nElkader, IA 52043\n(42.85350191...         None  
1972  201, E Washington St\nMount Pleasant, IA 52641...         None  

[1973 rows x 5 columns]

**Prompt 2 Answer**: 3061 Cub Foods / Sioux City

### Prompt 3

Among those sales whose category name is `IMPORTED VODKA`, if you sort by vendor in alphabetical order, what is the bottle price and number of bottles bought in the first 3 transactions?
> You should give us three sets of numbers - one for each transaction.

In [15]:
# Use a SQL query to find the answer to the above prompt.

sql = """

SELECT btl_price, bottle_qty
FROM sales
WHERE category_name = 'IMPORTED VODKA'
ORDER BY vendor DESC

"""

pd.read_sql_query(sql, engine)

btl_price  bottle_qty
0         $16.33           4
1         $16.33          12
2         $16.33          12
3         $16.33           3
4         $16.33           3
...          ...         ...
124670    $28.50          30
124671    $28.50           6
124672    $28.50           6
124673    $28.50          30
124674     $8.99          12

[124675 rows x 2 columns]

**Prompt 3 Answer**: 


First transaction:	$16.33	4
Second transaction:	$16.33	12
Third transaction:	$16.33	12


### Prompt 4

What is the total amount of liquor sold in February 2015?

In [17]:
# Use a SQL query to find the answer to the above prompt.

sql = """

SELECT SUM(total)
FROM sales
WHERE date BETWEEN '2-1-2015' AND '2-28-2015'

"""
pd.read_sql_query(sql, engine)

sum
0  21295350.55

**Prompt 4 Answer**:21295350.55

### Prompt 5

Among all transactions where `IMPORTED VODKA` was sold, which vendor has the most transactions? Report the total number of transactions, the number of bottles sold, and the total amount of revenue from these transactions. Rename the columns to make sure there isn't any confusion as to what they mean!

In [18]:
# Use a SQL query to find the answer to the above prompt.

sql = """

SELECT SUM(total), vendor, SUM(bottle_qty)
FROM sales
WHERE category_name = 'IMPORTED VODKA'


"""

pd.read_sql_query(sql, engine)

ProgrammingError: (psycopg2.errors.GroupingError) column "sales.vendor" must appear in the GROUP BY clause or be used in an aggregate function
LINE 3: SELECT SUM(total), vendor, SUM(bottle_qty)
                           ^

[SQL: 

SELECT SUM(total), vendor, SUM(bottle_qty)
FROM sales
WHERE category_name = 'IMPORTED VODKA'


]
(Background on this error at: http://sqlalche.me/e/13/f405)

**Prompt 5 Answer**: 

### Prompt 6

The reason we sell things is to make money. In sales, the term **markup** means the amount of extra money charged for a product over the cost to make that product. (You can think of the markup as the profit for that product.)

In Iowa, the law states the minimum price at which liquor may be sold.

Calculate the markup (name this `Markup`) that shows the difference between the `btl_price` and the `state_btl_cost`. 

> If you haven't subtracted columns together before, [this StackOverflow question](https://stackoverflow.com/questions/7536996/multiplying-two-columns-in-sql-server/7537059) may be helpful.

In [20]:
# Use a SQL query to find the answer to the above prompt.

sql = """

SELECT btl_price - state_btl_cost as Markup 
FROM sales



"""

pd.read_sql_query(sql, engine)

markup
0         $5.75
1         $5.75
2         $5.75
3         $8.58
4         $2.50
...         ...
3049908   $5.75
3049909  $10.00
3049910   $5.75
3049911   $5.75
3049912  $10.00

[3049913 rows x 1 columns]

### Prompt 7

Calculate the "Percentage Markup" by dividing markup by the state bottle cost. (Rather than the total profit per bottle, this will tell you how much stores are increasing their price over the state minimum.)

After calculating the "Percentage Markup," calculate the average percentage markup per **store**. Which store has the largest average markup, and what is that average markup?

In [23]:
# Use a SQL query to find the answer to the above prompt.

sql = """

SELECT store, AVG((btl_price - state_btl_cost)/state_btl_cost)/ store 
FROM sales

"""

pd.read_sql_query(sql, engine)

ProgrammingError: (psycopg2.errors.GroupingError) column "sales.store" must appear in the GROUP BY clause or be used in an aggregate function
LINE 3: SELECT store, AVG((btl_price - state_btl_cost)/state_btl_cos...
               ^

[SQL: 

SELECT store, AVG((btl_price - state_btl_cost)/state_btl_cost)/ store 
FROM sales

]
(Background on this error at: http://sqlalche.me/e/13/f405)

**Prompt 7 Answer**: 

### Prompt 8

Building off of your last query, retrieve the five stores with the highest average markup percentage. Along with the store IDs and average markup percentage, please return the minimum and maximum markup percentage for those stores, as well as the names and addresses of those store.

What are the names of the five stores?

In [ ]:
# Use a SQL query to find the answer to the above prompt.

sql = """

"""

pd.read_sql_query(sql, engine)

**Prompt 8 Answer**: Error with previous qn

### Prompt 9

Which nine counties sell the highest number of bottles of liquor per capita (per person)? Calculate the average bottle size of the bottles of liquor sold by these counties rounded to the nearest first decimal place. Of the nine counties selling the highest number of bottles of liquor per capita, which county sells (on average) the largest bottle size?

**_BONUS FUN_**: This question can (also) be solved with a SQL subquery - that is, a query that queries a query! If you'd like to give this a try, you can read more about it [here](https://www.dofactory.com/sql/subquery). Otherwise, keep an eye out for the solution key!

In [ ]:
# Use a SQL query to find the answer to the above prompt.

sql = """

"""

pd.read_sql_query(sql, engine)

**Prompt 9 Answer**:

### Prompt 10 (advanced but required!)

In the `sales` table, three columns are `btl_price`, `bottle_qty`, and `total`. The `total` for a transaction _should be_ the product of `btl_price` and `bottle_qty`. How many transactions have a value of `total` that is not equal to `btl_price` time `bottle_qty`?

> There will be a type error as well! We cannot compare type `money` to type `real`. We have to convert `total` to type `money` so that we can compare the two directly. [This link](https://docs.microsoft.com/en-us/sql/t-sql/functions/cast-and-convert-transact-sql?view=sql-server-ver15#syntax) may be helpful to you!

In [25]:
# Use a SQL query to find the answer to the above prompt.

sql = """

SELECT COUNT(*)
FROM sales
WHERE CAST(total AS money) <> (btl_price * bottle_qty);


"""

pd.read_sql_query(sql, engine)

count
0      0

**Prompt 10 Answer**: 0 such cases